File: 1.- Practica/p1.py
# Elabora un programa que resuelva este rompecabezas:
# "The bridge and torch problem"
# El enunciado del problema es el siguiente:

# Cuatro individuos llegan a un río de noche. Hay un puente estrecho, pero este solo soporta a dos personas a la vez. Los individuos 
# tienen una antorcha y, debido a que es de noche, deben utilizarla cuando cruzan el puente; por lo tanto, si cruzan dos personas, 
# uno debe volver atrás llevando la antorcha para que puedan cruzar los demás. El individuo A puede cruzar el puente en un minuto, 
# el individuo B en dos minutos, el individuo C en cinco minutos, y el individuo D en ocho minutos. Cuando dos individuos cruzan el 
# puente juntos, tardan lo que tarda el más lento de ellos.

# Tu programa deberá recibir una lista que represente el tiempo que tarda cada sujeto en cruzar el puente.
# La salida consiste en generar todos los movimientos de los sujetos, de forma que todos crucen el puente en el menor tiempo posible.

# [1,2,5,10] el óptimo es 17:
# (1,2) → 2; 1 ← 1; (5,10) → 10; 2 ← 2; (1,2) → 2. Total = 17.

#Librerias: 
# Librerias: 
import tkinter as tk
from itertools import combinations

# Variables globales
persona = [1, 2, 5, 10]
text_result = None  # lo usamos para mostrar resultados en Tkinter

def imprimirTitulo(): 
    return "\n Practica Uno: The bridge and torch problem \n"

def timeOptimo(persona):
    # Estado inicial
    inicial = (frozenset(persona), frozenset(), 0, 'izq')  
    soluciones = []
    mejorTiempo = [float('inf')]

    def dfs(izq, der, tiempo, antorcha, camino):
        if tiempo >= mejorTiempo[0]:
            return

        if not izq:
            if tiempo < mejorTiempo[0]:
                mejorTiempo[0] = tiempo
                soluciones.clear()
                soluciones.append((tiempo, camino[:]))
            elif tiempo == mejorTiempo[0]:
                soluciones.append((tiempo, camino[:]))
            return

        if antorcha == 'izq':
            for a, b in combinations(izq, 2):
                nuevoTiempo = tiempo + max(a, b)
                nuevoIzq = set(izq) - {a, b}
                nuevoDer = set(der) | {a, b}
                dfs(frozenset(nuevoIzq), frozenset(nuevoDer), nuevoTiempo, 'der',
                    camino + [f"{a},{b} → ({max(a,b)})"])
        else:
            for a in der:
                nuevoTiempo = tiempo + a
                nuevoIzq = set(izq) | {a}
                nuevoDer = set(der) - {a}
                dfs(frozenset(nuevoIzq), frozenset(nuevoDer), nuevoTiempo, 'izq',
                    camino + [f"{a} ← ({a})"])

    # Llamada inicial
    dfs(inicial[0], inicial[1], inicial[2], inicial[3], [])
    return soluciones

def imprimirResultados():
    soluciones = timeOptimo(persona)
    # limpiar antes de escribir
    text_result.delete("1.0", tk.END)  
    text_result.insert(tk.END, imprimirTitulo())
    text_result.insert(tk.END, f"Mejor tiempo: {soluciones[0][0]} minutos\n\n")
    text_result.insert(tk.END, "Una de las soluciones óptimas:\n")
    for pasoDado in soluciones[0][1]:
        text_result.insert(tk.END, f"  {pasoDado}\n")

def main():
    global text_result
    ventana = tk.Tk()
    ventana.title("The Bridge and Torch Problem")
    ventana.geometry("500x400")

    label_title = tk.Label(ventana, text="Practica Uno: The Bridge and Torch Problem", font=("Arial", 14, "bold"))
    label_title.pack(pady=10)

    btn_run = tk.Button(ventana, text="Resolver", command=imprimirResultados, 
                        font=("Arial", 12), bg="#FF0000", fg="white")
    btn_run.pack(pady=10)

    text_result = tk.Text(ventana, wrap="word", font=("Consolas", 11))
    text_result.pack(expand=True, fill="both", padx=10, pady=10)

    ventana.mainloop()

main()

File: 1.- Practica/java/P1.java
package p1;

import javax.swing.*;
import java.awt.Color;
import java.awt.Font;
import java.awt.BorderLayout;
import java.util.HashSet;
import java.util.Set;
import java.util.ArrayList;
import java.util.List;

public class P1 {

    private static final int[] persona = {1, 2, 5, 10};
    private static JTextArea textResult;

    // Función que imprime el título
    private static String imprimirTitulo() {
        return "\n Practica Uno: The bridge and torch problem \n";
    }

    // Función que encuentra el tiempo óptimo
    private static List<Solucion> timeOptimo(int[] persona) {
        Set<Integer> izqInicial = new HashSet<>();
        for (int p : persona) izqInicial.add(p);

        List<Solucion> soluciones = new ArrayList<>();
        int[] mejorTiempo = {Integer.MAX_VALUE};

        dfs(new HashSet<>(izqInicial), new HashSet<>(), 0, "izq", new ArrayList<>(), soluciones, mejorTiempo);

        return soluciones;
    }

    private static void dfs(Set<Integer> izq, Set<Integer> der, int tiempo, String antorcha,
                            List<String> camino, List<Solucion> soluciones, int[] mejorTiempo) {
        if (tiempo >= mejorTiempo[0]) return;

        if (izq.isEmpty()) {
            if (tiempo < mejorTiempo[0]) {
                mejorTiempo[0] = tiempo;
                soluciones.clear();
                soluciones.add(new Solucion(tiempo, new ArrayList<>(camino)));
            } else if (tiempo == mejorTiempo[0]) {
                soluciones.add(new Solucion(tiempo, new ArrayList<>(camino)));
            }
            return;
        }

        if (antorcha.equals("izq")) {
            List<Integer> listaIzq = new ArrayList<>(izq);
            for (int i = 0; i < listaIzq.size(); i++) {
                for (int j = i + 1; j < listaIzq.size(); j++) {
                    int a = listaIzq.get(i);
                    int b = listaIzq.get(j);

                    int nuevoTiempo = tiempo + Math.max(a, b);

                    Set<Integer> nuevoIzq = new HashSet<>(izq);
                    nuevoIzq.remove(a);
                    nuevoIzq.remove(b);

                    Set<Integer> nuevoDer = new HashSet<>(der);
                    nuevoDer.add(a);
                    nuevoDer.add(b);

                    List<String> nuevoCamino = new ArrayList<>(camino);
                    nuevoCamino.add(a + "," + b + " → (" + Math.max(a, b) + ")");

                    dfs(nuevoIzq, nuevoDer, nuevoTiempo, "der", nuevoCamino, soluciones, mejorTiempo);
                }
            }
        } else {
            for (int a : new ArrayList<>(der)) {
                int nuevoTiempo = tiempo + a;

                Set<Integer> nuevoIzq = new HashSet<>(izq);
                nuevoIzq.add(a);

                Set<Integer> nuevoDer = new HashSet<>(der);
                nuevoDer.remove(a);

                List<String> nuevoCamino = new ArrayList<>(camino);
                nuevoCamino.add(a + " ← (" + a + ")");

                dfs(nuevoIzq, nuevoDer, nuevoTiempo, "izq", nuevoCamino, soluciones, mejorTiempo);
            }
        }
    }

    // Clase para guardar soluciones
    private static class Solucion {
        int tiempo;
        List<String> pasos;

        Solucion(int tiempo, List<String> pasos) {
            this.tiempo = tiempo;
            this.pasos = pasos;
        }
    }

    // Función que imprime los resultados en el JTextArea
    private static void imprimirResultados() {
        List<Solucion> soluciones = timeOptimo(persona);

        textResult.setText(""); // limpiar antes de escribir
        textResult.append(imprimirTitulo());
        textResult.append("Mejor tiempo: " + soluciones.get(0).tiempo + " minutos\n\n");
        textResult.append("Una de las soluciones óptimas:\n");
        for (String paso : soluciones.get(0).pasos) {
            textResult.append("  " + paso + "\n");
        }
    }

    // Ventana principal
    public static void main(String[] args) {
        JFrame ventana = new JFrame("The Bridge and Torch Problem");
        ventana.setSize(500, 400);
        ventana.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        ventana.setLayout(new BorderLayout());

        JLabel labelTitle = new JLabel("Practica Uno: The Bridge and Torch Problem", JLabel.CENTER);
        labelTitle.setFont(new Font("Arial", Font.BOLD, 14));
        ventana.add(labelTitle, BorderLayout.NORTH);

        JButton btnRun = new JButton("Resolver");
        btnRun.setFont(new Font("Arial", Font.PLAIN, 12));
        btnRun.setBackground(Color.RED);
        btnRun.setForeground(Color.WHITE);
        btnRun.addActionListener(e -> imprimirResultados());
        ventana.add(btnRun, BorderLayout.SOUTH);

        textResult = new JTextArea();
        textResult.setFont(new Font("Consolas", Font.PLAIN, 11));
        textResult.setWrapStyleWord(true);
        textResult.setLineWrap(true);

        JScrollPane scroll = new JScrollPane(textResult);
        ventana.add(scroll, BorderLayout.CENTER);

        ventana.setVisible(true);
    }
}
